src/app/clusters/service-area-server/service-area-server.h

```diff
@@ -0,0 +1,751 @@
+     * @note Selected locations are normally only set through the SelectLocations command.
+     * @note This function is intended primarily for testing.
+    */
+    bool AddSelectedLocation(uint32_t & aSelectedLocation);
+
+    /**
+     * @brief Get the list of selected locations
+     * @param aSelectedLocations empty vector to be filled with location Id's
+    */
+    void GetSelectedLocations(std::vector<uint32_t> & aSelectedLocations);
+    
+    /**
+     * @brief Is the location in the selected locations list?
+     * @param aLocationId id of the location
+     * @return true if the location identified by Id is in the selected locations list
+     */ 
+    bool IsSelectedLocation(uint32_t aLocationId);
+
+    /**
+     * @brief remove any selected locations which are no longer supported location (locationId not in supported locations)
+     * @return true if any locations were removed from selected locations.
+     * @note caller is responsible for change notification
+     */
+    bool PruneSelectedLocations();
+
+    /**
+     * @brief remove a location from the selected locations list
+     * @return true if the location was removed from selected locations.
+     * @note caller is responsible for change notification
+     */
+    bool DeleteSelectedLocation(uint32_t aLocationId);
+
+    /**
+     * @brief Clear the Selected Locations list
+     * @note caller is responsible for change notification
+     * @return the Selected Locations list changed (i.e. it previously had entries)
+     */
+    bool ClearSelectedLocationsList();
+
+
+    //*************************************************************************
+    // Current Location manipulators
+
+    /**
+     * @brief Get the Current Location 
+     * @return The current location
+     */
+    DataModel::Nullable<uint32_t> GetCurrentLocation();
+
+    /**
+     * @brief Set the current location
+     * @param aCurrentLocation where the device is currently. Must be a supported location or NULL.
+     * @return true if the current location is set
+     * @note if current location is set to null, estimated end time will be set to null
+     */ 
+    bool SetCurrentLocation(const DataModel::Nullable<uint32_t> & aCurrentLocation);
+
+
+    //*************************************************************************
+    // Estimated End Time manipulators
+
+    /**
+     * @brief Get the estimated end time
+     * @return The estimated epoch time in seconds when operation at the location indicated by the CurrentLocation attribute will be completed, in seconds.
+     */
+    DataModel::Nullable<uint32_t> GetEstimatedEndTime();
+
+    /**
+     * @brief Set the estimated end time
+     * @param aEstimatedEndTime The estimated epoch time in seconds when operation at the location indicated by the CurrentLocation attribute will be completed, in seconds.
+     * @return true if attribute is set
+     * @note change notification is generated (!!) if appropriate per cluster requirements
```

Comment:
Most of the simple attribute functions don't handle notifications. Some of the more complex ones do.
Extensive commenting in the doxygen, but could still be confusing.

In this case because notifications depend so much on the old and new values, it made sense to add the notification code in this simpler function.